src/LFHCAL_geo.cpp

```diff
@@ -0,0 +1,915 @@
+      continue;
+    }
+
+    std::string delimiter = " ";
+    size_t posC = 0;
+    std::string token;
+    while ((posC = mtrx_values.find(delimiter)) != std::string::npos) {
+      token = mtrx_values.substr(0, posC);
+      aptr->push_back(atof(token.c_str()));
+      mtrx_values.erase(0, posC + delimiter.length());
+    }
+    aptr->push_back(atof(mtrx_values.c_str()));
+  }
+
+  if (xpos8M.size() != ypos8M.size() || xpos8M.size() != zpos8M.size()){
+    std::cout << xpos8M.size() << "\t" <<ypos8M.size() <<  "\t" << zpos8M.size() << std::endl;
+    std::cout <<  "idiot you can't count" << std::endl;
+  } else {
+    for (int e = 0; e < (int)xpos8M.size(); e++){
+      std::cout <<  "LFHCAL placing 8M module: " << e << "/"<< (int)xpos8M.size() << "\t" << xpos8M[e] << "\t" << ypos8M[e] << "\t" << zpos8M[e]<< std::endl;
+      Assembly  eightMassembly = createEightMModule ( desc, eightM_params, slice_Params, moduleID, length, sens);
+
+      // Placing modules in world volume
+      auto tr8M = Transform3D(Position(pos.x()-xpos8M[e]*dd4hep::mm-0.5*eightM_params.mod_width, pos.y()-ypos8M[e]*dd4hep::mm-0.5*eightM_params.mod_height, pos.z() +zpos8M[e]*dd4hep::mm + length / 2.));
+      phv = motherVol.placeVolume(eightMassembly, tr8M);
+      moduleID++;
+    }
+  }
+
+  std::vector<double> xpos4M;
+  std::vector<double> ypos4M;
+  std::vector<double> zpos4M;
+
+  for(xml_coll_t i(handle, _Unicode(fourmodulepositions)); i; ++i){
+    xml_comp_t  x_mtrx = i;
+
+    std::string   mtrx_name       = getAttrOrDefault<std::string>(x_mtrx, _Unicode(name), " ");
+    std::string   mtrx_values     = getAttrOrDefault<std::string>(x_mtrx, _Unicode(values), " ");
+
+    std::vector<double> *aptr = NULL;
+
+    if(mtrx_name == "xpos")
+      aptr = &xpos4M;
+    else if(mtrx_name == "ypos")
+      aptr = &ypos4M;
+    else if(mtrx_name == "zpos")
+      aptr = &zpos4M;
+    else{
+      printout(WARNING, "LFHCAL", "unknown <fourmodulepositions> data!");
+      continue;
+    }
+
+    std::string delimiter = " ";
+    size_t posC = 0;
+    std::string token;
+    while ((posC = mtrx_values.find(delimiter)) != std::string::npos) {
+      token = mtrx_values.substr(0, posC);
+      aptr->push_back(atof(token.c_str()));
+      mtrx_values.erase(0, posC + delimiter.length());
+    }
+    aptr->push_back(atof(mtrx_values.c_str()));
+  }
+
+  // create 4M modules
+  if (xpos4M.size() != ypos4M.size() || xpos4M.size() != zpos4M.size()){
+    std::cout << xpos4M.size() << "\t" <<ypos4M.size() <<  "\t" << zpos4M.size() << std::endl;
+    std::cout <<  "idiot you can't count" << std::endl;
+  } else {
+    for (int f = 0; f < (int)xpos4M.size(); f++){
+      std::cout <<  "LFHCAL placing 4M module: " << f << "/"<< (int)xpos4M.size() << "\t" << xpos4M[f] << "\t" << ypos4M[f] << "\t" << zpos4M[f]<< std::endl;
+      Assembly  fourMassembly = createFourMModule ( desc, fourM_params, slice_Params,  moduleID, length, sens);
```

Comment:
If these assemblies are identical, you can create their logical volumes just once and use `placeVolume` to put copies of it in different locations as needed. Same goes for all the nested pieces down the hierarchy.

Reply:
I tried that, but unfortunately that didn't work as I needed to assign each active volume the correct IDs

Reply:
I'm submitting a similar change for EcalBarrel right now. Here is how it's done:
https://github.com/eic/epic/pull/293/commits/8cedd81efcfedff3d18c3abc90a1d9a93d490269#diff-3eed1d7a563d681a4ff46f655d092d270bb844260f59462fa5aecdbdc845a904L282
https://github.com/eic/epic/pull/293/commits/8cedd81efcfedff3d18c3abc90a1d9a93d490269#diff-3eed1d7a563d681a4ff46f655d092d270bb844260f59462fa5aecdbdc845a904R434

Reply:
https://github.com/eic/epic/pull/293/files?diff=unified&w=1#diff-3eed1d7a563d681a4ff46f655d092d270bb844260f59462fa5aecdbdc845a904 (this should hide the whitespace changes)

Reply:
> I'm submitting a similar change for EcalBarrel right now. Here is how it's done: [8cedd81#diff-3eed1d7a563d681a4ff46f655d092d270bb844260f59462fa5aecdbdc845a904L282](https://github.com/eic/epic/commit/8cedd81efcfedff3d18c3abc90a1d9a93d490269#diff-3eed1d7a563d681a4ff46f655d092d270bb844260f59462fa5aecdbdc845a904L282) [8cedd81#diff-3eed1d7a563d681a4ff46f655d092d270bb844260f59462fa5aecdbdc845a904R434](https://github.com/eic/epic/commit/8cedd81efcfedff3d18c3abc90a1d9a93d490269#diff-3eed1d7a563d681a4ff46f655d092d270bb844260f59462fa5aecdbdc845a904R434)

That would work if one of my modules would be the equivalent of a readout unit, unfortunately this is not the case.  One 8M modules has right now 8 * 65 readout channels.

Reply:
It is possible to further define values sub-segmentation for sub volumes. This is done in many other detectors, including SciGlass linked above.

Reply:
@veprbl Can you resolve this conversation if this looks good to you?

Reply:
No, this has not been implemented. Not addressing this will result in higher memory use (and slower simulation?). If that's okay, we could merge without it.

Reply:
Nevermind. See new commits. Looks better, resolved.

Reply:
Could we assess the time it takes to simulate per event, before and after, and the memory it takes after?